File: solutions/java/twelve-days/1/src/main/java/TwelveDays.java
import java.util.Arrays;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
class TwelveDays {
    private final String suffix = "th";
    private final String[] DAYS = {"first", "second", "third", "four", "fif", "six", "seven", "eigh", "nin", "ten", "eleven", "twelf"}
    , GIFTS = {"twelve Drummers Drumming", "eleven Pipers Piping", "ten Lords-a-Leaping", "nine Ladies Dancing", "eight Maids-a-Milking", "seven Swans-a-Swimming", "six Geese-a-Laying", "five Gold Rings", "four Calling Birds", "three French Hens", "two Turtle Doves"};
    private final String FIRST_GIFT = "a Partridge in a Pear Tree";
    
    String verse(int verseNumber) {
        return String.format("On the %s day of Christmas my true love gave to me: %s.\n"
            , DAYS[verseNumber - 1] + (verseNumber < 4 ? "" : suffix)
            , verseNumber == 1 ? FIRST_GIFT : Arrays.stream(GIFTS).skip(GIFTS.length - verseNumber + 1).collect(Collectors.joining(", ")) + ", and " + FIRST_GIFT);
    }

    String verses(int startVerse, int endVerse) {
        return IntStream.rangeClosed(startVerse, endVerse).mapToObj(this::verse).collect(Collectors.joining("\n"));
    }
    
    String sing() {
        return verses(1, DAYS.length);
    }
}

File: solutions/python/twelve-days/1/twelve_days.py
def recite(start_verse: int, end_verse: int) -> list[str]:
    return list(map(get_verse, range(start_verse, end_verse + 1)))


def get_verse(number: int) -> str:
    GIFTS = ("a Partridge in a Pear Tree", "two Turtle Doves", "three French Hens",
             "four Calling Birds", "five Gold Rings", "six Geese-a-Laying",
             "seven Swans-a-Swimming", "eight Maids-a-Milking", "nine Ladies Dancing",
             "ten Lords-a-Leaping", "eleven Pipers Piping", "twelve Drummers Drumming")
    
    return "On the {} day of Christmas my true love gave to me: {}.".format(
        ("first", "second", "third", "fourth", "fifth", "sixth", "seventh", "eighth", "ninth", "tenth", "eleventh", "twelfth")[number - 1],
        GIFTS[0] if number == 1 else ", ".join(GIFTS[number-1:0:-1]) + ", and " + GIFTS[0]
    )
